### PR TITLE
Fix multi-session sync bug, introduce resync logic and improve sync config

### PR DESCRIPTION
**Pull request by mycognosist**
- Fix bugs in sync attempt state-tracking which resulted in multiple concurrent attempts
- Introduce `SyncConfiguration` type (with internal `ResyncConfiguration`) for fine-grained control over sync sessions
- Add resync ability

### CODE REVIEW

File: p2panda-net/src/sync/manager.rs [hunk 1]
```diff
@@ -176,9 +185,29 @@ where
...
+                    if let Err(err) = self.schedule_attempt(sync_attempt).await {
+                        // The attempt will fail if the sync queue is full, indicating that a high
+                        // volume of sync sessions are underway. In that case, we drop the attempt
+                        // completely. Another attempt will be scheduled when the next announcement of
+                        // this peer-topic combination is received from the network-wide gossip
+                        // overlay.
+                        error!("failed to schedule sync attempt: {}", err)
+                    }
+                }
+                _ = attempt_resync_interval.tick() => {
```
Comment (mycognosist) [inline p2panda-net/src/sync/manager.rs:199]: I wonder if there's a better solution for this resync behaviour...

Right now we're using a simple `VecDeque`: every 5 seconds we pull an attempt from the queue and schedule it if the previous attempt was completed more than 60 seconds ago. If not, we push it to the back of the queue.

Is a 5 second interval frequent enough? What if we end up with 50 or 100 attempts in the queue? Do we need a solution that offers reattempt timing guarantees (e.g. reattempt will be made more than 60 seconds after the previous successful attempt, but not more than 70 seconds after)?

File: p2panda-net/src/network.rs [hunk 2]
```diff
@@ -176,9 +178,15 @@ where
...
     ///
     /// If a sync protocol is provided, a sync session will be completed with any known peers with
-    /// whom we share topics of interest.
-    pub fn sync(mut self, protocol: impl for<'a> SyncProtocol<'a, T> + 'static) -> Self {
+    /// whom we share topics of interest. The resync flag determines whether sync should only be
+    /// completed once or repeated at regular intervals.
+    pub fn sync(
+        mut self,
+        protocol: impl for<'a> SyncProtocol<'a, T> + 'static,
+        resync: bool,
```
Comment (adzialocha) [inline p2panda-net/src/network.rs:186]: Would be cool to pass a whole `SyncConfiguration` struct here, so we can add further options in the future. There's already a few others which are currently hard-coded which could live here.

Comment (mycognosist) [reply p2panda-net/src/network.rs:186]: Ya, good idea!

File: p2panda-net/src/sync/manager.rs (hunk 1 above)
Comment (adzialocha) [reply p2panda-net/src/sync/manager.rs]: The logic looks fine as-is to me right now.

"First ever sync attempts" should always have higher priority I believe and "> 1th attempts" lower priority, but I think this is currently given with the two separate "priority" channels (`inbox` vs. `resync_queue`).

It gets hard to reason about timings when there's a higher priority channel involved with unknown number of items, so I think it's fine to say we can't define the upper timing limit, only the minimum.

**Review by adzialocha — APPROVED** [review]

**Pull request merged**